srml/contracts/src/wasm/runtime.rs

```diff
@@ -835,6 +835,29 @@ define_env!(Env, <E: Ext>,
 		ctx.ext.block_number().encode_to(&mut ctx.scratch_buf);
 		Ok(())
 	},
+
+	// Retrieve the value at the under the given key from the **runtime** storage and return 0.
+	// If there is no entry under the given key then this function will return 1 and
+	// clear the scratch buffer.
+	//
+	// - key_ptr: the pointer into the linear memory where the requested value is placed.
+	// - key_len: the length of the key in bytes.
```

Comment:
The first sentence doesn't make sense grammatically.

Reply:
Also I would suggest the following docs:

```
/// Retrieve the value at the given key from the **runtime** storage and return 0.
///
/// # Error
///
/// - If there is no entry at the given key return 1.
/// - A runtime might restrict access to a certain subset of keys. If the supplied
///   key is not part of that restricted whitelist of keys; return 1.
///
/// # Parameters
///
/// - `key_ptr`: The pointer into the contiguous memory where the requested value is placed.
/// - `key_len`: The length of the key in number of bytes.
```

Reply:
These docs include your idea with the whitelisted keys. I think we should really implement this as it is a really useful safety feature. We might be inclined to return a different value from `1` though in this case.

Reply:
I generally would also like an `# Errors` and `# Parameters` section for all the `ext_` function docs because it looks clearer and also is more similar to what normal Rust docs look like.

Reply:
ah, I meant to remove `at the` part, thanks.

as per whitelist let's discuss it on the meeting tmrw.

wrt doc sections: yeah sounds good. As with other feedback items that do not relate to this particular PR, I'd move them as a follow up. (Otherwise we end up with mixed styles, which I consider to not be not worth it even though it would improve the situation of one particular item)

Reply:
> really useful safety feature

How, exactly? All storage is necessarily public and open. A proof can always be given about the last block's state, or attested to in the current block by the block author. So *at best* you're delaying the ability for a contract to "read" storage by one block assuming that the block author doesn't want to facilitate things.

Reply:
Hm ... so the thoughts were as followed: A contract could read storage from a runtime entry that is arbitrarily large and thus requires an arbitrarily amount of time (resources) to query as far as I understood. So the idea by @pepyakin and me was that whitelisting keys (entries) by the runtime would fix this problem.

Reply:
@gavofyork 

To give more context: for substrate contracts we limit the maximal size of stored values by some limit (~16Kb). Up to this number, the size of the value for storage loads don't influence the run time significantly and the idea that we would apply flat rate for the loads.

However, if you allow loading around the whole storage, this assumption might not hold and then the value size aspect kicks in. This is really annoying since we don't have a way to find out the size of the storage value to compute and charge gas for that. This can be abused if you have contract calls that devote a bit of gas and then make a big load (e.g. `:code`).

Reply:
Hmm, yeah, this does seem like a problem. Another issue is that even if the value you are loading isn't too big, one of it's ancestor nodes in the trie could have a huge attached value (eg. doing a lookup for `:code:nothing-here`). As long as storage lookups require a trie traversal, this would also incur a large performance penalty. I don't see any options to `Get` in RocksDB that allows the caller to limit the size of the value.